File: frontend/src/apis/customer/RechargeCustomerBalanceApi.ts
import type { CustomerId, DeliveryAppState, RechargeBalanceRequest } from '@/objects/core/SharedObjects'
import { postNormalizedDeliveryState } from '@/system/api/DeliveryStateClient'
import { buildApiPath1, defineJsonPostApi1, routeSegment } from '@/system/api/TypedApiDefinitions'

export const rechargeCustomerBalanceApiDefinition = defineJsonPostApi1<CustomerId, RechargeBalanceRequest, DeliveryAppState>(
    [routeSegment('api'), routeSegment('delivery'), routeSegment('customers')],
    [routeSegment('recharge')],
  )

export function rechargeCustomerBalance(
  customerId: CustomerId,
  payload: RechargeBalanceRequest,
) {
  return postNormalizedDeliveryState(buildApiPath1(rechargeCustomerBalanceApiDefinition, customerId), payload)
}

File: backend/src/main/scala/services/customer/api/RechargeCustomerBalanceApi.scala
package services.customer.api

import domain.shared.given

import cats.effect.IO
import services.customer.utils.rechargeCustomerBalance
import domain.customer.RechargeBalanceRequest
import domain.shared.{CustomerId, DeliveryAppState, UserRole}
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import system.api.*
import system.app.*

val rechargeCustomerBalanceApi: FixedMethodApi1[CustomerId, DeliveryAppState] =
  jsonPostApi1[CustomerId, RechargeBalanceRequest, DeliveryAppState](
    List(routeSegment("api"), routeSegment("delivery"), routeSegment("customers")),
    List(routeSegment("recharge")),
  )
